Block level boxes with block-step-size other than none should establish an independent formatting context.

sammygill opened the pull request:
#### 423f7933224ac4609944bf6e5a18932f2c337b37
<pre>
Block level boxes with block-step-size other than none should establish an independent formatting context.
<a href="https://bugs.webkit.org/show_bug.cgi?id=252358">https://bugs.webkit.org/show_bug.cgi?id=252358</a>
rdar://105519556

Reviewed by Tim Nguyen.

In order to determine whether this property would apply to a box we
first need to check its display type. According to the table from
css-display-3 (<a href="https://www.w3.org/TR/css-display-3/#the-display-properties)">https://www.w3.org/TR/css-display-3/#the-display-properties)</a>,
block level boxes are generated from the following display types:
block, flow-root, list-item, flex, grid, block ruby, and table. Of these,
flow-root, flex, grid, and table already establish an independent
formatting context. So we need to make sure that a new block formatting
context is generated for the remaining, which is what the tests cover.

Inside RenderBox::establishesBlockFormattingContext and
RenderElement::establishesIndependentFormattingContext we will call
into RenderElement::isBlockLevelElementWithBlockStepSizing which will
return true if the box has a block level display type and if it has
a valid value for block-step-sizing (other than none).

div {
    width: 50px;
    height: 100px;
    background-color: green;
}
.floating {
    float: left;
}
.block-step-size {
    block-step-size: 1px;
}
&lt;/style&gt;
&lt;div class=&quot;floating&quot;&gt;&lt;/div&gt;
&lt;div class=&quot;block-step-size&quot;&gt;&lt;/div&gt;

Since both of the conditions hold for the second div it will establish a
new block formatting context and will become aware of the float so that
it does not get positioned over it.

Spec reference: <a href="https://drafts.csswg.org/css-rhythm/#block-step-size">https://drafts.csswg.org/css-rhythm/#block-step-size</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-rhythm/block-step-size-establishes-independent-formatting-context-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-rhythm/block-step-size-establishes-independent-formatting-context-list-item-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-rhythm/block-step-size-establishes-independent-formatting-context-list-item.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-rhythm/block-step-size-establishes-independent-formatting-context.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-rhythm/block-step-size-none-does-not-establish-indepdendent-formatting-context-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-rhythm/block-step-size-none-does-not-establish-indepdendent-formatting-context.html: Added.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::establishesBlockFormattingContext const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::establishesIndependentFormattingContext const):
(WebCore::RenderElement::isBlockLevelElementWithBlockStepSizing const):
* Source/WebCore/rendering/RenderElement.h:

Canonical link: <a href="https://commits.webkit.org/261065@main">https://commits.webkit.org/261065@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f2bcfd49552aa47279d5ab5eaec22120b4a1239

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110351 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19441 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42966 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1742 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119288 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114299 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20884 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10608 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102597 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116095 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15556 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98752 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43772 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97517 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30417 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85627 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12109 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31755 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12700 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8721 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18065 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51370 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7663 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14545 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -2123,7 +2123,12 @@ bool RenderElement::createsNewFormattingContext() const
+    return isFloatingOrOutOfFlowPositioned() || hasPotentiallyScrollableOverflow() || style().containsLayout() || paintContainmentApplies() || isBlockLevelElementWithBlockStepSizing();
+}
+
+bool RenderElement::isBlockLevelElementWithBlockStepSizing() const
```
Not sure we need `isBlockLevelElement` in the name, the spec makes it clear that block-step-size only applies to block-level containers.

In fact I wonder if this should be a method on `RenderStyle`, something like `RenderStyle::effectiveBlockStepSize()` that returns `none` when it's inline level. 

@alanbaradlay Any opinions here?

alanbaradlay replied:
please check "9.2.1 Block-level elements and block boxes" for proper naming
[visuren](https://www.w3.org/TR/CSS22/visuren.html) or just simply look at LFC (all these names have been properly resolved in LFC long time ago)

alanbaradlay replied:
@anttijk may have some opinions on the RenderStyle part.

alanbaradlay replied:
the problem with effective* pattern in general is that it's highly error prone as people could easily overlook it and go with the not-effective (what does that even mean) call (e.g. blockStepSize() in this case)...also you need to know what it actually means to decided whether you need to call this or the other one. Not a big fan.

sammygill replied:
I just ended up removing the function and extracted out the logic into the method I was using it at. If I really end up needing this in multiple locations we can consider adding it back with perhaps a different name or approach

alanbaradlay commented on Source/WebCore/rendering/RenderElement.h:
```diff
@@ -311,6 +311,7 @@ class RenderElement : public RenderObject {
 
     virtual bool establishesIndependentFormattingContext() const;
     bool createsNewFormattingContext() const;
+    bool isBlockLevelElementWithBlockStepSizing() const;
```
I am not sure if this should be function on RenderElement. We don't usually have a "BoxType + CSS property" kind of highly specific APIs -especially if it's only used in couple of places.

sammygill replied:
Not necessarily. It was mostly helpful when reading through all of the conditions in establishesIndependentFormattingContext. Tim's suggestion would be just as good or it could even be an lambda

alanbaradlay commented on LayoutTests/imported/w3c/web-platform-tests/css/css-rhythm/block-step-size-none-does-not-establish-indepdendent-formatting-context.html:
```diff
@@ -0,0 +1,30 @@
+<div class="floating"></div>
+<div class="block-step-size"></div>
+</body>
+</html>
```
what exactly are we trying to test here? (the more tests, the better, but independent formatting context does not make a box to be a float avoider, BFC does, so maybe just stick to testing BFC behavior? -also, relatively positioned float?)

sammygill replied:
Yes, the goal here was to test that the block container with block-step-size created a BFC. My understanding is that when a block container establishes a formatting context it is a new BFC (along with maybe an IFC), so I think that's what should be happening here and I can make sure it is more clear in the test, but let me know if my logic is wrong. It's from me interpreting this: https://drafts.csswg.org/css-display/#independent-formatting-context

alanbaradlay replied:
yea, the BFC part sounds correct (but that's not what the test case's assert says: 

> name="assert" content="block-step-size values other than none on block level elements cause the box to establish an independent formatting context"

sammygill replied:
Fixed the test case asserts and file names to make it clear it was testing BFC

alanbaradlay replied:
don't wanna be too nitpicy, but I am still seeing "independent formatting context" in some of those files.

sammygill replied:
No thank you for catching this!

sammygill committed:
Block level boxes with block-step-size other than none should establish an independent formatting context.
https://bugs.webkit.org/show_bug.cgi?id=252358
rdar://105519556

Reviewed by Tim Nguyen.

In order to determine whether this property would apply to a box we
first need to check its display type. According to the table from
css-display-3 (https://www.w3.org/TR/css-display-3/#the-display-properties),
block level boxes are generated from the following display types:
block, flow-root, list-item, flex, grid, block ruby, and table. Of these,
flow-root, flex, grid, and table already establish an independent
formatting context. So we need to make sure that a new block formatting
context is generated for the remaining, which is what the tests cover.

Inside RenderBox::establishesBlockFormattingContext and
RenderElement::establishesIndependentFormattingContext we will call
into RenderElement::isBlockLevelElementWithBlockStepSizing which will
return true if the box has a block level display type and if it has
a valid value for block-step-sizing (other than none).

div {
    width: 50px;
    height: 100px;
    background-color: green;
}
.floating {
    float: left;
}
.block-step-size {
    block-step-size: 1px;
}
</style>
<div class="floating"></div>
<div class="block-step-size"></div>

Since both of the conditions hold for the second div it will establish a
new block formatting context and will become aware of the float so that
it does not get positioned over it.

Spec reference: https://drafts.csswg.org/css-rhythm/#block-step-size

* LayoutTests/imported/w3c/web-platform-tests/css/css-rhythm/block-step-size-establishes-independent-formatting-context-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-rhythm/block-step-size-establishes-independent-formatting-context-list-item-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-rhythm/block-step-size-establishes-independent-formatting-context-list-item.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-rhythm/block-step-size-establishes-independent-formatting-context.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-rhythm/block-step-size-none-does-not-establish-indepdendent-formatting-context-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-rhythm/block-step-size-none-does-not-establish-indepdendent-formatting-context.html: Added.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::establishesBlockFormattingContext const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::establishesIndependentFormattingContext const):
(WebCore::RenderElement::isBlockLevelElementWithBlockStepSizing const):
* Source/WebCore/rendering/RenderElement.h:

Canonical link: https://commits.webkit.org/261065@main
